src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,452 @@
+  rw not_exists at h',
+
+  -- Then `(n)` is a proper ideal in `R`.
+  let I := ideal.span ({n} : set R),
+  have hI : (1 : R) ∉ I :=
+  begin
+    by_contradiction one_mem_I,
+    cases ideal.mem_span_singleton'.mp one_mem_I with x hx,
+    exact absurd hx (h' x),
+  end,
+
+  -- by assumption `char_zero R⧸(n)` so `n = 0`.
+  have h_char_zero : char_zero (R ⧸ I) := (h.elim) I ((ideal.ne_top_iff_one I).mpr hI),
+  have n_zero : (n : ℕ) = 0 :=
+  begin
+    apply h_char_zero.cast_injective,
+    rw ←map_nat_cast (ideal.quotient.mk I),
+    rw [nat.cast_zero, ideal.quotient.eq_zero_iff_mem, ideal.mem_span_singleton'],
+    use 1,
+    simp,
+  end,
+  exact absurd n_zero (ne_of_gt n.property),
+end
+
+/-- Internal: Not intended to be used outside this local construction. -/
+noncomputable instance equal_char_zero.pnat_has_coe_units
+  [fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I))] : has_coe_t ℕ+ Rˣ :=
+⟨λn, (equal_char_zero.pnat_coe_is_unit R n).unit⟩
+
+/-- Internal: Not intended to be used outside this local construction. -/
+lemma equal_char_zero.pnat_coe_units_eq_one [fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I))] :
+  ((1 : ℕ+) : Rˣ) = 1 :=
+begin
+  apply units.ext,
+  rw units.coe_one,
+  change ((equal_char_zero.pnat_coe_is_unit R 1).unit : R) = 1,
+  rw is_unit.unit_spec (equal_char_zero.pnat_coe_is_unit R 1),
+  rw [coe_coe, pnat.one_coe, nat.cast_one],
+end
+
+/-- Internal: Not intended to be used outside this local construction. -/
+lemma equal_char_zero.pnat_coe_units_coe_eq_coe
+  [fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I))] (n : ℕ+) :
+  ((n : Rˣ) : R) = ↑n :=
+begin
+  change ((equal_char_zero.pnat_coe_is_unit R n).unit : R) = ↑n,
+  simp only [is_unit.unit_spec],
+end
+
+/--
+Equal characteristic implies `ℚ`-algebra.
+-/
+noncomputable def equal_char_zero_to_Q_algebra (h : ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) :
+  algebra ℚ R :=
+begin
+  -- turning `h` into a `fact` since we used it that way in the helper lemmas above.
+  haveI : fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) := by {rw fact_iff, exact h},
+
+  apply ring_hom.to_algebra,
+  exact
+  { to_fun := λ x, x.num /ₚ ↑(x.pnat_denom),
+    map_zero' := by simp [divp],
+    map_one' := by simp [equal_char_zero.pnat_coe_units_eq_one],
+    map_mul' :=
+    begin
+      intros a b,
+      field_simp,
+      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
+      convert_to (↑((a * b).num * (a.denom) * (b.denom)) : R) = _,
+      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
+        ring },
+      rw rat.mul_num_denom' a b,
+      simp
+    end,
+    map_add' :=
+    begin
+      intros a b,
+      field_simp,
+      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
+      convert_to (↑((a + b).num * a.denom * b.denom) : R)  = _,
+      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
+        ring },
+      rw rat.add_num_denom' a b,
+      simp
+    end }
+end
```

Comment:
This is not so important, but usually for constructing data you don't want to stay in tactic mode: that can really produce unexpected definitions. So the idiom for adding an instance locally is `by haveI; exact (term)`.
```suggestion
by haveI : fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) := ⟨h⟩; exact
ring_hom.to_algebra
  { to_fun := λ x, x.num /ₚ ↑(x.pnat_denom),
    map_zero' := by simp [divp],
    map_one' := by simp [equal_char_zero.pnat_coe_units_eq_one],
    map_mul' :=
    begin
      intros a b,
      field_simp,
      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
      transitivity (↑((a * b).num * (a.denom) * (b.denom)) : R),
      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
        ring },
      rw rat.mul_num_denom' a b,
      simp
    end,
    map_add' :=
    begin
      intros a b,
      field_simp,
      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
      transitivity (↑((a + b).num * a.denom * b.denom) : R),
      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
        ring },
      rw rat.add_num_denom' a b,
      simp
    end }
```